docs/configuration.md
---
layout: doc
title: Configuration
---

### Pomegranate takes 1 object for its configuration.

#### From a file? Environment vars? How you provide it is up to you.

```javascript
var options = {
    port: 8080,
    address: "0.0.0.0",
    session: false
}
```

<a name="module_ServerOptions"></a>
## ServerOptions
Controls server operation.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| port | <code>number</code> | <code>8080</code> | Server listening port. |
| address | <code>string</code> | <code>&quot;0.0.0.0&quot;</code> | Server listening address. |
| templating | <code>string</code> | <code>&quot;hbs&quot;</code> | Installed with Pomegranate [hbs, jade] |
| renderErrors | <code>string</code> | <code>false</code> | If true renders ./views/errors/404 ./views/errors/500 on error, else returns json. |



<a name="module_PathOptions"></a>
## PathOptions
Controls Pomegranate file locations.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| basePath | <code>string</code> | <code>&quot;./&quot;</code> | This should generally be set to __dirname. |
| staticFiles | <code>string</code> | <code>&quot;./public&quot;</code> | Static files directory. Default is relative to the requiring file. |
| views | <code>string</code> | <code>&quot;./views&quot;</code> | View files directory. Default is relative to the requiring file. |
| partials | <code>string</code> | <code>&quot;./views/partials&quot;</code> | Directory containing handlebars partials. Only for templating = hbs |
| routes | <code>string</code> | <code>&quot;./routes&quot;</code> | Route files directory. Default is relative to the requiring file. |
| models | <code>string</code> | <code>&quot;./models&quot;</code> | Model files directory. Default is relative to the requiring file. |
| controllers | <code>string</code> | <code>&quot;./controllers&quot;</code> | Controller files directory. Default is relative to the requiring file. |
| dependencies | <code>string</code> | <code>&quot;./dependencies&quot;</code> | Dependency files directory. Default is relative to the requiring file. |



<a name="module_SessionOptions"></a>
## SessionOptions
Controls Server sessions and session storage.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| session | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | Session Configuration. See [Express-session](https://github.com/expressjs/session) |
| session.secret | <code>string</code> | <code>&quot;process.exit()&quot;</code> | No default, Pomegranate will exit if not set. |
| session.domain | <code>string</code> | <code>false</code> | No default, Domain to store cookie on. |
| session.name | <code>string</code> | <code>&quot;pomegranate.sid&quot;</code> | Session cookie name. |
| session.ttl | <code>number</code> | <code>60 * 60 * 24 * 7</code> | Expiration time for session data. In seconds. |
| session.resave | <code>boolean</code> | <code>false</code> | Save session on every request, even if unmodified. |
| session.rolling | <code>boolean</code> | <code>false</code> | Reset the session cookie expiration on every request. |
| session.saveUninitialized | <code>boolean</code> | <code>false</code> | Save new but unmodified session data. |



<a name="module_RedisOptions"></a>
## RedisOptions
Controls Redis connection settings.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| redis | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | Redis connection details. See [Node Redis](https://github.com/mranney/node_redis) |
| redis.host | <code>string</code> | <code>&quot;localhost&quot;</code> | Redis server address. |
| redis.port | <code>number</code> | <code>6379</code> | Redis server port. |
| redis.password | <code>string</code> | <code>null</code> | Redis auth password. |



<a name="module_SQLOptions"></a>
## SQLOptions
Controls SQL connection settings [Postgres, MySQL, MsSQL, MariaDB].

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| sql | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | SQL connection details. See [Sequelize](https://github.com/sequelize/sequelize) |
| sql.database | <code>string</code> | <code>&quot;no default&quot;</code> | SQL server database. |
| sql.username | <code>string</code> | <code>&quot;no default&quot;</code> | SQL User. |
| sql.password | <code>string</code> | <code>&quot;no default&quot;</code> | SQL User password. |
| sql.host | <code>string</code> | <code>&quot;localhost&quot;</code> | SQL server address. |
| sql.port | <code>number</code> | <code>no default</code> | Sql Server Port. |
| sql.dialect | <code>string</code> | <code>&quot;sqlite&quot;</code> | SQL server dialect 'mysql','mariadb','sqlite','postgres','mssql' |
| sql.storage | <code>string</code> | <code>&quot;no default&quot;</code> | SQLITE only, path to db file. |
| sql.logging | <code>boolean</code> | <code>false</code> | LOG SQL queries |
| sql.persist | <code>boolean</code> | <code>true</code> | Destroys all model databases and recreates on startup. |



<a name="module_CouchOptions"></a>
## CouchOptions
Controls CouchDB connection settings.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| couch | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | CouchDb connection details. See [nano](https://github.com/dscape/nano) |
| couch.url | <code>string</code> | <code>&quot;no default&quot;</code> | CouchDB connection url. |
| couch.prefix | <code>string</code> | <code>&quot;pomegranate&quot;</code> | Used to prefix database names, to support multitenancy. |
| couch.persist | <code>boolean</code> | <code>true</code> | Destroys all model databases and recreates on startup. |



_gh-pages/docs/configuration.md
---
layout: doc
title: Configuration
---

### Pomegranate takes 1 object for its configuration.

#### From a file? Environment vars? How you provide it is up to you.

```javascript
var options = {
    port: 8080,
    address: "0.0.0.0",
    session: false
}
```

<a name="module_ServerOptions"></a>
## ServerOptions
Controls server operation.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| port | <code>number</code> | <code>8080</code> | Server listening port. |
| address | <code>string</code> | <code>&quot;0.0.0.0&quot;</code> | Server listening address. |
| templating | <code>string</code> | <code>&quot;hbs&quot;</code> | Installed with Pomegranate [hbs, jade] |
| renderErrors | <code>string</code> | <code>false</code> | If true renders ./views/errors/404 ./views/errors/500 on error, else returns json. |



<a name="module_PathOptions"></a>
## PathOptions
Controls Pomegranate file locations.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| basePath | <code>string</code> | <code>&quot;./&quot;</code> | This should generally be set to __dirname. |
| staticFiles | <code>string</code> | <code>&quot;./public&quot;</code> | Static files directory. Default is relative to the requiring file. |
| views | <code>string</code> | <code>&quot;./views&quot;</code> | View files directory. Default is relative to the requiring file. |
| partials | <code>string</code> | <code>&quot;./views/partials&quot;</code> | Directory containing handlebars partials. Only for templating = hbs |
| routes | <code>string</code> | <code>&quot;./routes&quot;</code> | Route files directory. Default is relative to the requiring file. |
| models | <code>string</code> | <code>&quot;./models&quot;</code> | Model files directory. Default is relative to the requiring file. |
| controllers | <code>string</code> | <code>&quot;./controllers&quot;</code> | Controller files directory. Default is relative to the requiring file. |



<a name="module_SessionOptions"></a>
## SessionOptions
Controls Server sessions and session storage.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| session | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | Session Configuration. See [Express-session](https://github.com/expressjs/session) |
| session.secret | <code>string</code> | <code>&quot;process.exit()&quot;</code> | No default, Pomegranate will exit if not set. |
| session.name | <code>string</code> | <code>&quot;pomegranate.sid&quot;</code> | Session cookie name. |
| session.ttl | <code>number</code> | <code>60 * 60 * 24 * 7</code> | Expiration time for session data. In seconds. |
| session.resave | <code>boolean</code> | <code>false</code> | Save session on every request, even if unmodified. |
| session.rolling | <code>boolean</code> | <code>false</code> | Reset the session cookie expiration on every request. |
| session.saveUninitialized | <code>boolean</code> | <code>false</code> | Save new but unmodified session data. |



<a name="module_RedisOptions"></a>
## RedisOptions
Controls Redis connection settings.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| redis | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | Redis connection details. See [Node Redis](https://github.com/mranney/node_redis) |
| redis.host | <code>string</code> | <code>&quot;localhost&quot;</code> | Redis server address. |
| redis.port | <code>number</code> | <code>6379</code> | Redis server port. |
| redis.password | <code>string</code> | <code>null</code> | Redis auth password. |



<a name="module_SQLOptions"></a>
## SQLOptions
Controls SQL connection settings [Postgres, MySQL, MsSQL, MariaDB].

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| sql | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | SQL connection details. See [Sequelize](https://github.com/sequelize/sequelize) |
| sql.database | <code>string</code> | <code>&quot;no default&quot;</code> | SQL server database. |
| sql.username | <code>string</code> | <code>&quot;no default&quot;</code> | SQL User. |
| sql.password | <code>string</code> | <code>&quot;no default&quot;</code> | SQL User password. |
| sql.host | <code>string</code> | <code>&quot;localhost&quot;</code> | SQL server address. |
| sql.port | <code>number</code> | <code>no default</code> | Sql Server Port. |
| sql.dialect | <code>string</code> | <code>&quot;sqlite&quot;</code> | SQL server dialect 'mysql','mariadb','sqlite','postgres','mssql' |
| sql.storage | <code>string</code> | <code>&quot;no default&quot;</code> | SQLITE only, path to db file. |
| sql.logging | <code>boolean</code> | <code>false</code> | LOG SQL queries |
| sql.persist | <code>boolean</code> | <code>true</code> | Destroys all model databases and recreates on startup. |



<a name="module_CouchOptions"></a>
## CouchOptions
Controls CouchDB connection settings.

**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| couch | <code>object</code> &#124; <code>boolean</code> | <code>false</code> | CouchDb connection details. See [nano](https://github.com/dscape/nano) |
| couch.url | <code>string</code> | <code>&quot;no default&quot;</code> | CouchDB connection url. |
| couch.prefix | <code>string</code> | <code>&quot;pomegranate&quot;</code> | Used to prefix database names, to support multitenancy. |
| couch.persist | <code>boolean</code> | <code>true</code> | Destroys all model databases and recreates on startup. |



docs/changes.md
---
layout: doc
title: Changes
---

1.0.1 / 2015-07-22
==================

  * 1.0.1
  * Added change log, added docs for routing and route files.
  * Update package.json files entry
  * Added licence file

1.0.0 / 2015-07-21
==================
  * 1.0.0
  * Added documentation site.
  * Expose the default middleware stack and allow it to be altered.
  * Route definitions now derived from file path.

0.3.0 / 2015-07-16
==================

  * 0.3.0
  * Added path-parse polyfill for 10.x tests
  * Path routes based on file structure, rather than returned object.

0.2.7 / 2015-07-03
==================

  * 0.2.7
  * Added no op function check to Pomegranate#start

0.2.6 / 2015-07-01
==================

  * 0.2.6
  * Added config option to render errors vs sending JSON response

0.2.5 / 2015-07-01
==================

  * 0.2.5
  * Finalized favicon path handling

0.2.4 / 2015-07-01
==================

  * 0.2.4
  * Fixed pathing issue when running under a process manager or from parent cli app

0.2.3 / 2015-07-01
==================

  * 0.2.3
  * Add additional check for valid favicon file

0.2.2 / 2015-07-01
==================

  * 0.2.2
  * Expose handlebars registerPartials via config setting

0.2.1 / 2015-07-01
==================

  * 0.2.1
  * Moved controller loader to async api.
  * Updated hiredis to v0.4.0, was killing builds on iojs v2.x.x

0.2.0 / 2015-06-18
==================

  * 0.2.0
  * mirrored MagnumDI API in Pomegranate#addDependency

0.1.0 / 2015-06-16
==================

  * 0.1.0
  * Update Dependencies

0.0.7 / 2015-06-15
==================

  * 0.0.7
  * Added Controllers to the mix

0.0.6 / 2015-05-23
==================

  * 0.0.6
  * CouchLoader now handles multiple view definitions

0.0.5 / 2015-05-22
==================

  * 0.0.5
  * Finished Couch Model loading, still needs better error handlers
  * Point example link to correct location

0.0.4 / 2015-05-19
==================

  * 0.0.4
  * Adding support for couch

0.0.3 / 2015-05-13
==================

  * 0.0.3
  * Code reorganization, moved DI module to its own project, added assets.
    * Grouped bare files into directories.
    * The Dependency Injection Module has been moved into its own project at PaperElectron/Magnum-DI
    * Added favicon file and source.

0.0.2 / 2015-05-07
==================

  * 0.0.2
  * IoC


0.0.1 / 2015-04-30
==================

  * 0.0.1
  * Name change
  * Initial commit

lib/Configuration/Config.js
/**
 * @file Config
 * @author Jim Bulkowski <jim.b@paperelectron.com>
 * @project Pomegranate
 * @license MIT {@link http://opensource.org/licenses/MIT}
 */
var path = require('path');

module.exports = function(opts) {

  var mainPath;

  /**
   *  Get a reliable route to our base application directory.
   */
  if(opts.basePath){
    mainPath = opts.basePath
  } else {
    mainPath = path.dirname(require.main.filename);
  }

  var staticFiles = path.join(mainPath, 'public');
  var defaultViews = path.join(mainPath, 'views');
  var defaultPartials = path.join(defaultViews, 'partials');
  var defaultRoutes = path.join(mainPath, 'routes');
  var defaultModels = path.join(mainPath, 'models');
  var defaultControllers = path.join(mainPath, 'controllers');
  var defaultDependencies = path.join(mainPath, 'dependencies');
  /**
   * Controls server operation.
   * @module ServerOptions
   * @property {number} [port = 8080] Server listening port.
   * @property {string} [address = 0.0.0.0] Server listening address.
   * @property {string} [templating = hbs] Installed with Pomegranate [hbs, jade]
   * @property {string} [renderErrors = false] If true renders ./views/errors/404 ./views/errors/500 on error, else returns json.
   */

  /**
   * Controls Pomegranate file locations.
   * @module PathOptions
   * @property {string} [basePath = ./] This should generally be set to __dirname.
   * @property {string} [staticFiles = ./public] Static files directory. Default is relative to the requiring file.
   * @property {string} [views = ./views] View files directory. Default is relative to the requiring file.
   * @property {string} [partials = ./views/partials] Directory containing handlebars partials. Only for templating = hbs
   * @property {string} [routes = ./routes] Route files directory. Default is relative to the requiring file.
   * @property {string} [models = ./models] Model files directory. Default is relative to the requiring file.
   * @property {string} [controllers = ./controllers] Controller files directory. Default is relative to the requiring file.
   * @property {string} [dependencies = ./dependencies] Dependency files directory. Default is relative to the requiring file.
   */

  /**
   * Controls Server sessions and session storage.
   * @module SessionOptions
   * @property {object|boolean} [session = false] Session Configuration.
   * See [Express-session]{@link https://github.com/expressjs/session}
   * @property {string} [session.secret = process.exit()] No default, Pomegranate will exit if not set.
   * @property {string} [session.domain = false] No default, Domain to store cookie on.
   * @property {string} [session.name = pomegranate.sid] Session cookie name.
   * @property {number} [session.ttl = 60 * 60 * 24 * 7] Expiration time for session data. In seconds.
   * @property {boolean} [session.resave = false] Save session on every request, even if unmodified.
   * @property {boolean} [session.rolling = false] Reset the session cookie expiration on every request.
   * @property {boolean} [session.saveUninitialized = false] Save new but unmodified session data.
   */

  /**
   * Controls Redis connection settings.
   * @module RedisOptions
   * @property {object|boolean} [redis = false] Redis connection details.
   * See [Node Redis]{@link https://github.com/mranney/node_redis}
   * @property {string} [redis.host = localhost] Redis server address.
   * @property {number} [redis.port = 6379] Redis server port.
   * @property {string} [redis.password = null] Redis auth password.
   */

  /**
   * Controls SQL connection settings [Postgres, MySQL, MsSQL, MariaDB].
   * @module SQLOptions
   * @property {object|boolean} [sql = false] SQL connection details.
   * See [Sequelize]{@link https://github.com/sequelize/sequelize}
   * @property {string} [sql.database = no default] SQL server database.
   * @property {string} [sql.username = no default] SQL User.
   * @property {string} [sql.password = no default] SQL User password.
   * @property {string} [sql.host = localhost] SQL server address.
   * @property {number} [sql.port = no default] Sql Server Port.
   * @property {string} [sql.dialect = sqlite] SQL server dialect 'mysql','mariadb','sqlite','postgres','mssql'
   * @property {string} [sql.storage = no default] SQLITE only, path to db file.
   * @property {boolean} [sql.logging = false] LOG SQL queries
   * @property {boolean} [sql.persist = true] Destroys all model databases and recreates on startup.
   */

  /**
   * Controls CouchDB connection settings.
   * @module CouchOptions
   * @property {object|boolean} [couch = false] CouchDb connection details.
   * See [nano]{@link https://github.com/dscape/nano}
   * @property {string} [couch.url = no default] CouchDB connection url.
   * @property {string} [couch.prefix = pomegranate] Used to prefix database names, to support multitenancy.
   * @property {boolean} [couch.persist = true] Destroys all model databases and recreates on startup.
   */

  return {
    port: opts.port || 8080,
    address: opts.address || '0.0.0.0',
    templating: opts.templating || 'hbs',
    renderErrors: opts.renderErrors || false,
    views: opts.views || defaultViews,
    partials: opts.partials || defaultPartials,
    staticFiles: opts.staticFiles || staticFiles,
    routes: opts.routes || defaultRoutes,
    models: opts.models || defaultModels,
    controllers: opts.controllers || defaultControllers,
    dependencies: opts.dependencies || defaultDependencies,
    session: opts.session || false,
    redis: opts.redis || false,
    sql: opts.sql || false,
    couch: opts.couch || false
  };
};
